process: Add code to warnings, assign codes to deprecations

jasnell opened the pull request:
##### Checklist
- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

process, lib

##### Description of change

As discussed at the Node.js Interactive Collaboration Summit in Austin last week, this PR adds a few new capabilities that should make it easier to make dealing with runtime deprecation warnings in Node.js easier.

1. A static identifier code can be assigned to all process warnings. These are specified as either a new optional argument to `process.emitWarning()` or as a `code` property on an `Error` object passed into `process.emitWarning()`. Within the `process.on('warning')` event handler, these can be accessed via the `code` property on the `warning` object.

2. A new `--redirect-warnings=file` command line argument and matching `NODE_REDIRECT_WARNINGS=file` environment variable is added to allow process warning output to be printed to the given `file` as opposed to dumping to `stderr`. This will allow, for instance, npm and other applications spawning node process instances to capture warning output with less noise to the console.

3. Every runtime and documentation-only deprecation in Node.js is assigned a static identifier code.

4. A new deprecations.md file is added to the API documentation that lists each of the runtime and documentation-only deprecation codes assigned. For future work (perhaps a good first contribution for new Node.js contributors), it would be helpful to expand the details in deprecations.md to include links to the relevant sections in other parts of the API documentation as well as expanded descriptions of why the APIs are deprecated and what the alternatives are. /cc @nodejs/documentation 

*Note*: This PR is currently semver-major because it adds improved type checking to the input on `process.emitWarning()` to ensure that both the warning name and the identifier code are strings.
If this PR is backported to older Node.js release streams, the backport can be landed as a semver-minor *if and only if* the additional type checking on the `name` argument is removed.

/cc @ashleygwilliams

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5176/

addaleax commented on doc/api/cli.md:
```diff
@@ -121,6 +121,14 @@ added: v6.0.0
+-->
+
+Write process warnings to the given file instead of printing to stderr but
+only if the file does not already exist.
```
Why “only if the file does not already exist”? I’m not sure I’d expect that, and it sounds a bit like that might interfere with programmatic use of the API.

Is this about multiple node processes sharing the same env var?

jasnell replied:
I'm not entirely sure about the `only if the file does not already exist` part either. For now I have it this way just to ensure that something important isn't accidentally overwritten.

Fishrock123 replied:
Maybe we should just append?

jasnell replied:
Just appending works also. Doing so is pretty safe when dealing with multiple node processes since the warning output includes the pid

addaleax commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,491 @@
+`OutgoingMessage.prototype.flushHeaders()` instead.
+
+<a id="DEP00002"></a>
+### DEP00002: require('_linklist')
```
This should probably be `\_linklist` (including the `\`)

addaleax commented on doc/api/process.md:
```diff
@@ -706,14 +706,15 @@ console.log(process.env.test);
 * `warning` {String | Error} The warning to emit.
 * `name` {String} When `warning` is a String, `name` is the name to use
   for the warning. Default: `Warning`.
+* `code` {String} A unique identifier for the warning.
```
I’m not sure the semantic difference between `name` and `code` is obvious here?

Fishrock123 replied:
Name may be better as "type"?

addaleax replied:
> Name may be better as "type"?

Remember that `.name` is inherited from the `Error` classes – I don’t think we can change that.

I’m also not worried that much about the naming, a bit of clarification might be all that’s needed

jasnell replied:
The argument can be called `type` with `name` still used on the error object. I'm not overly concerned about the difference there.

addaleax commented on lib/buffer.js:
```diff
@@ -63,14 +63,14 @@ function alignPool() {
+ * The Buffer() construtor is deprecated in documentation and should not be
+ * used moving forward. Rather, developers should use one of the three new
+ * factory APIs: Buffer.from(), Buffer.allocUnsafe() or Buffer.alloc() based on
+ * their specific needs. There is no hard deprecation because of the extent to
```
nit: `Runtime deprecation`? ;)

addaleax commented on lib/buffer.js:
```diff
@@ -63,14 +63,14 @@ function alignPool() {
+ * used moving forward. Rather, developers should use one of the three new
+ * factory APIs: Buffer.from(), Buffer.allocUnsafe() or Buffer.alloc() based on
+ * their specific needs. There is no hard deprecation because of the extent to
+ * which the Buffer constructor is used in the ecosystem currently -- a hard
```
(ditto)

addaleax commented on lib/internal/process/warning.js:
```diff
@@ -1,9 +1,38 @@
 
+var fd;
+function acquireFd() {
+  if (!fd) {
```
Nit: `fd === undefined`

addaleax commented on lib/internal/process/warning.js:
```diff
@@ -13,26 +42,44 @@ function setupProcessWarnings() {
+    }
+    if (code && typeof code !== 'string')
+      throw new TypeError('`code` must be a String');
+    if (name && typeof name !== 'string')
```
nit: `name !== undefined` and `code !== undefined`

addaleax commented on lib/internal/util.js:
```diff
@@ -51,11 +51,18 @@ exports._deprecate = function(fn, msg) {
     return fn;
   }
 
+  if (code && typeof code !== 'string')
```
(ditto)

addaleax commented on src/node_config.cc:
```diff
@@ -44,6 +45,16 @@ void InitConfig(Local<Object> target,
   if (config_preserve_symlinks)
     READONLY_BOOLEAN_PROPERTY("preserveSymlinks");
+
+  if (config_warning_file != NULL) {
```
nit: `nullptr`

jasnell commented:
@addaleax @Fishrock123 ... updated to address the nits. Updated the `--redirect-warnings` to append and renamed the `emitWarning()` `name` argument to `type`

jasnell commented:
I disagree. The method already has the option of passing in a Error object
as the first parameter. Switching to an options object would just add
confusion, I think.

On Sun, Dec 4, 2016 at 4:44 PM silverwind <notifications@github.com> wrote:

> *@silverwind* commented on this pull request.
> ------------------------------
>
> In doc/api/process.md
> <https://github.com/nodejs/node/pull/10116#pullrequestreview-11307207>:
>
> > @@ -706,14 +706,15 @@ console.log(process.env.test);
>  // => 1
>  ```
>
> -## process.emitWarning(warning[, name][, ctor])
> +## process.emitWarning(warning[, type[, code]][, ctor])
>
> I think a options object is in order here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/10116#pullrequestreview-11307207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAa2ec-j0kNP2A0JnuBeKN3ejbShHNY-ks5rE15ugaJpZM4LDsqo>
> .
>

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,491 @@
+
+Type: Runtime
+
+The `crypto.Credentials` class is deprecated. Please Uuse `tls.SecureContext`
```
use

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,491 @@
+Type: Runtime
+
+The `constants` module has been deprecated. When requiring access to constants
+relevant to specific Node.js built in modules, developers should instead refer
```
built-in

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,491 @@
+
+Type: Runtime
+
+The `fs.read()` legacy String interface is deprecated. Use the Buffer API as
```
We explicitly ask the readers to refer documentation, it would be better if we had the link here.

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,491 @@
+
+Type: Documentation-only
+
+The `domain` module is considered to be deprecated and should not be used.
```
Considered to be? We deprecated it, right?

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,491 @@
+
+Type: Documentation-only
+
+The `require.extensions` API has been deprecated.
```
property or API

thefourtheye commented on lib/constants.js:
```diff
@@ -1,8 +1,9 @@
 'use strict';
 
-// This module is soft-deprecated. Users should be directed towards using
+// This module is deprecate in docs. Users should be directed towards using
```
doc-deprecated?

thefourtheye commented on lib/internal/process/warning.js:
```diff
@@ -13,26 +42,44 @@ function setupProcessWarnings() {
+        if (warning.code) {
+          output(`${prefix}[${warning.code}] ${toString.apply(warning)}`);
+        } else {
+          output(`${prefix}${toString.apply(warning)}`);
```
Should there be a space after `${prefix}`?

jasnell replied:
No, the ${prefix} includes a space already

thefourtheye commented on src/node_config.cc:
```diff
@@ -44,6 +45,16 @@ void InitConfig(Local<Object> target,
+  if (config_warning_file != nullptr) {
+    target->DefineOwnProperty(env->context(),
+                              OneByteString(env->isolate(), "warningFile"),
+                              String::NewFromUtf8(env->isolate(),
```
Can we split them into multiple variables?

thefourtheye commented on test/parallel/test-process-redirect-warnings-env.js:
```diff
@@ -0,0 +1,29 @@
+fork(warnmod, {env: {NODE_REDIRECT_WARNINGS: warnpath}})
+  .on('exit', common.mustCall(() => {
+    fs.stat(warnpath, common.mustCall((err, stat) => {
+      assert(!err);
```
`assert.ifError`

silverwind commented:
> Switching to an options object would just add confusion, I think.

It's more of a general adivce to design APIs in a extendable way. I think that once a function takes 3 or more arguments, it gets to a point where the user has to lookup the arguments on less-frequently used functions.

Fishrock123 commented:
What do we do about things we have to un-deprecate? Just remove them from the doc? Or mark them as no longer deprecated?

jasnell commented:
I'd leave them in the doc and label them as undeprecated. The codes should
increase monotonically so leaving them in the doc will avoid collisions

On Mon, Dec 5, 2016 at 7:10 AM Jeremiah Senkpiel <notifications@github.com>
wrote:

> What do we do about things we have to un-deprecate? Just remove them from
> the doc? Or mark them as no longer deprecated?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/10116#issuecomment-264878487>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAa2eWrYaP8LINcM1uUgqt8HFsZHXYIaks5rFClXgaJpZM4LDsqo>
> .
>

jasnell commented:
@silverwind ... I don't disagree, but as I said, this API already provides the option of passing in an `Error` object as the first argument. Given that, I don't see significant value in providing an options object

jasnell commented:
@thefourtheye ... updated to address your feedback. I squashed and force pushed to keep things organized.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/5211/

jasnell commented:
@nodejs/ctc ... thoughts on this?

trevnorris commented:
@jasnell there a way to automatically track the next deprecation code? feel it could be error prone or frustrating to grep all the current ones to find the next.

jasnell commented:
There might be, but we need to make sure they stay static once assigned.
Best bet would be to check the deprecations.md and assign the next highest,
then make sure that doc always gets updated with a new deprecation

On Fri, Dec 9, 2016 at 2:11 PM Trevor Norris <notifications@github.com>
wrote:

> @jasnell <https://github.com/jasnell> there a way to automatically track
> the next deprecation code? feel it could be error prone or frustrating to
> grep all the current ones to find the next.
>
>
>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/10116#issuecomment-266137238>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAa2efCdwqCpSxTV1JESmkYNkFF7MQHwks5rGdIEgaJpZM4LDsqo>
> .
>

trevnorris commented:
@jasnell ah. missed `deprecations.md` b/c github refrained from loading it. okay, seems reasonable enough.

targos commented on lib/internal/process/warning.js:
```diff
@@ -1,9 +1,38 @@
+  if (typeof config.warningFile === 'string') {
+    try {
+      const fs = require('fs');
+      fs.appendFileSync(acquireFd(), `${message}\n`);
```
This will print a stack trace if `--trace-sync-io` is used. Are we OK with it?

jasnell replied:
Good point. Switching the fs.appendFile should not be a problem

jasnell commented:
@targos .. ok, switched over to using async to avoid the `--trace-sync-io` issue. Good thing, given that `--trace-sync-io` uses emitWarning using it here caused a recursion problem. That should be fixed now. Also made a couple of other improvements along the way. PTAL

evanlucas commented on lib/internal/process/warning.js:
```diff
@@ -1,9 +1,74 @@
+
+function lazyFs() {
+  if (!fs)
+    fs = require('fs');
```
Is this worth lazy loading? This will block and could happen after startup.

ChALkeR replied:
Does this affect `--trace-sync-io` output?

jasnell replied:
Depends on what you mean. This uses async fs methods so in that sense no, but since `--trace-sync-io` uses process.emitWarning, then using `--redirect-warnings` does redirect `--trace-sync-io` output to the warning file.

evanlucas commented on doc/node.1:
```diff
@@ -207,6 +211,11 @@ when outputting to a TTY on platforms which support async stdio.
 
+.TP
+.BR NODE_REDIRECT_WARNINGS=\fIfile\fR
+Write process warnings to the given file instead of printing to stderr.
```
Does this have to be an absolute path? Or does relative work?

jasnell replied:
ditto

evanlucas commented on doc/node.1:
```diff
@@ -112,6 +112,10 @@ Silence all process warnings (including deprecations).
 
 .TP
+.BR \-\-redirect\-warnings=\fIfile\fR
+Write process warnings to the given file instead of printing to stderr.
```
Does this have to be an absolute path? Or does relative work?

jasnell replied:
relative works.

jasnell commented:
@nodejs/ctc ... can I get some additional review and signoff on this please?

addaleax commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+# Deprecated APIs
+
+Node.js will deprecate APIs when either: (a) use of the API is considered to be
```
nit: I’d phrase it as `may deprecate`, otherwise it sounds like all APIs considered “unsafe” will eventually be deprecated, and we know that’s something we may not be able to guarantee.

addaleax commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+## Un-deprecation
+
+From time-to-time, the Node.js Core Technical Committee (CTC) may decide to
+reverse the decision to deprecate an API. Such action may happen in either a
```
I’d drop the mention of the CTC. It’s an implementation detail and it’s ultimately irrelevant who makes these decisions; how about just switching to passive voice, i.e. `From time to time, the deprecation of an API may be reversed.`?

addaleax commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+## List of Deprecated APIs
+
+<a id="DEP00001"></a>
+### DEP00001: OutgoingMessage.prototype.flush
```
Could you add a `http.` in front of the name to indicate what module this is coming from?

addaleax commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+
+Type: Runtime
+
+The [`crypto.createCredentials()`][] API is deprecated. Please Use
```
typo: lower-case `use`

addaleax commented on src/node.cc:
```diff
@@ -3472,6 +3475,7 @@ static void PrintHelp() {
          "       node debug script.js [arguments] \n"
          "\n"
          "Options:\n"
+<<<<<<< HEAD
```
um… ;)

jasnell replied:
grr... sigh

addaleax commented on src/node.cc:
```diff
@@ -3485,6 +3489,9 @@ static void PrintHelp() {
          "  --trace-warnings         show stack traces on process warnings\n"
+         "  --redirect-warnings=path\n"
+         "                        write process warnings to the given file\n"
+         "                        path rather than to stderr\n"
```
If it helps shorten this, I think you can just say `--redirect-warnings=path` and `write process warnings to path rather than to stderr`.

mscdex replied:
Also the description indentation is off a bit.

mscdex commented on src/node.cc:
```diff
@@ -4171,6 +4180,10 @@ void Init(int* argc,
     config_preserve_symlinks = (*preserve_symlinks == '1');
   }
 
+  if (auto redirect_warnings = secure_getenv("NODE_REDIRECT_WARNINGS")) {
```
Should this be included in the help text as well?

jasnell commented:
@addaleax @mscdex updated! Thank you!

mscdex commented on lib/internal/process/warning.js:
```diff
@@ -13,26 +78,44 @@ function setupProcessWarnings() {
+      code = undefined;
     }
+    if (code !== undefined && typeof code !== 'string')
+      throw new TypeError('`code` must be a String');
```
I thought we were using single quotes around parameter names instead of backticks in error messages?

jasnell replied:
Either way works for me. I'm also good with no quotes at all.

jasnell commented on lib/internal/process/warning.js:
```diff
@@ -44,6 +127,6 @@ function setupProcessWarnings() {
         throw warning;
     }
-    process.nextTick(() => process.emit('warning', warning));
+    process.nextTick(doEmitWarning(warning));
```
Since I was in here working, I took the opportunity to make a couple of additional cleanups that I'd been sitting on.

ChALkeR commented:
@jasnell Re: codes — do those really need five digits? Four would be enough for adding a new deprecation each day for 27 years, given that those are sequential.

With `DEP00001`, it's harder to tell how many zeros are there at the first sight, if the code is being manually typed and not copy-pasted.

ChALkeR commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+* [`Buffer.from(arrayBuffer[, byteOffset[, length]])`][from_arraybuffer] - Create a `Buffer`
+  that wraps the given `arrayBuffer`.
+* [`Buffer.from(buffer)`][] - Create a `Buffer` that copies `buffer`.
+* [`Buffer.from(string[, encoding])`][from_string_encoding] - Create a `Buffer` that copies
```
The documentation here needs updating as #10236 landed.

ChALkeR commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+# Deprecated APIs
```
This list misses `fs.SyncWriteStream`, see #6749.

jasnell commented:
Perhaps not, but the work that I am doing with introducing error codes is also looking at using five digits and I'm wanting to maintain (a) consistency and (b) guaranteed future proofing. Five digits may be a bit much but I see no reason to make it a smaller number.

ChALkeR commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,567 @@
+
+Type: Documentation-only
+
+The [`util._extend()`][] API has been deprecated.
```
This (and other) deprecation descriptions should probably be more explicit in terms of mentioning the fact that «deprecated» API in question was never documented to begin with. The documentation for `util._extend` first appeared in https://github.com/nodejs/node/pull/4903, in a PR that deprecated it.

Perhaps something «Was never part of the documented public API» or something like that will do here and in other places where applicable.

jasnell replied:
Yes, I agree, but this additional detail can (and should) be handled in a separate PR.

ChALkeR commented:
@jasnell I suppose that an error and a deprecation per day for 27 years is future-proof enough, isn't it?

The codes are needed for several reasons, those being:
 * programmatic checks and/or grouping/counts 
 * shorter unique identifiers that someone could google on.

For the second case — I expect that at least some percentage of people would manually type those, for various reasons. For that usecase, `DEP0005` is more preferrable than `DEP00005` as it's easier to read at a glance.

As for the error codes — I don't think those need five digits, too =).

I understand that the reasoning behind not wanting to add more digits in the future is that we don't want `DEP11111` to match with a regex for `DEP1111`, but I don't think we will ever get over 9999 of those in the foreseable future (once again, that's more than 27 * 365).

jasnell commented:
Ok ;-) ... I'll remove a digit ;-)

jasnell commented:
Ok, digit removed and missing deprecation added!

thefourtheye commented on lib/buffer.js:
```diff
@@ -64,14 +64,14 @@ function alignPool() {
+ * their specific needs. There is no runtime deprecation because of the extent
+ * to which the Buffer constructor is used in the ecosystem currently -- a
+ * runtime deprecation would introduce too much breakage at this time. It's not
+ * likely that the Buffer constructors would ever actually be removed.
```
Should we keep this line?

jasnell replied:
In this PR, yes.

thefourtheye commented on lib/fs.js:
```diff
@@ -2021,6 +2020,7 @@ WriteStream.prototype.destroySoon = WriteStream.prototype.end;
 // todo(jasnell): "Docs-only" deprecation for now. This was never documented
 // so there's no documentation to modify. In the future, add a runtime
 // deprecation.
+// Deprecation ID: DEP0061
```
`DEP0016`

thefourtheye commented on lib/internal/bootstrap_node.js:
```diff
@@ -204,7 +204,7 @@
       const get = util.deprecate(function() {
         return this;
-      }, `'${name}' is deprecated, use 'global'`);
+      }, `'${name}' is deprecated, use 'global'`, 'DEP0016');
```
If the warning in `fs.js` is actually `DEP0016`, then this and whatever follows this have to be changed.

thefourtheye commented on lib/internal/process/warning.js:
```diff
@@ -1,9 +1,74 @@
+var fs;
+var fd;
+var acquiringFd = false;
+function nop() {}
```
Can this be `const` to avoid accidental reassignment/declaration?

jasnell replied:
Could be but not convinced it's necessary

thefourtheye commented on lib/internal/process/warning.js:
```diff
@@ -1,9 +1,74 @@
+  return function(err, fd) {
+    if (err)
+      return writeOut(message);
+    lazyFs().appendFile(fd, `${message}\n`, nop);
```
Since we ignore the errors, should we at-least document that this is just a best-effort functionality?

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Runtime
+
+The `_writableState.buffer` property is deprecated. Use
+The `_writableState.getBuffer()` method instead.
```
`the`

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Runtime
+
+Within the [`child_process`][] module's `spawn()`, `fork()`, and `exec()`
+methods, The `options.customFds` option is deprecated. The `options.stdio`
```
the `options....`

jasnell replied:
Not sure what you're suggesting here

thefourtheye replied:
The - capital letter, in the middle of the sentence

jasnell replied:
Fixed!

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Runtime
+
+When using the `cluster` module, the [`worker.suicide`][] property has been
```
"In `cluster` module, the ...", wouldn't be better?

thefourtheye commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+<a id="DEP0008"></a>
+### DEP0008: require('constants')
+
+Type: Runtime
```
This is actually doc-only deprecation. Are we changing this?

jasnell replied:
No, just a typo. This PR does not change any of the existing deprecations

jasnell commented:
Updated to address nits

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5630/

jasnell commented:
Failures in FreeBSD are unrelated.

Trott commented:
Just cleaned up some old test fixtures that were still running on one of the FreeBSD hosts. That should hopefully fix the issue on CI. Let's try again. FreeBSD CI: https://ci.nodejs.org/job/node-test-commit-freebsd/6140/

Trott commented:
Uh, yeah, cleared out a FreeBSD 11 host, but there was also a FreeBSD 10 host. Let's try again, again: https://ci.nodejs.org/job/node-test-commit-freebsd/6142/

jasnell commented:
CI is green. @nodejs/ctc ... can I get some sign offs?

Trott commented on test/parallel/test-process-redirect-warnings.js:
```diff
@@ -0,0 +1,29 @@
+
+fork(warnmod, {execArgv: [`--redirect-warnings=${warnpath}`]})
+  .on('exit', common.mustCall(() => {
+    fs.stat(warnpath, common.mustCall((err, stat) => {
```
Is the `stat()` call necessary? Why not just let `fs.readFile()` fail if the file isn't there? Reduces the nesting by one. If I'm not mistaken, the "stat() followed by read()" is kind of an anti-pattern that we should probably discourage because there's nothing (in theory) preventing another process from deleting the file between the `stat()` and the `read()`.

Maybe I'm missing something here, but it seems this could just use `fs.readFileSync()` here and be a one- or two-liner that's probably easier to understand.

jasnell replied:
Fixed

Trott commented on test/parallel/test-process-redirect-warnings-env.js:
```diff
@@ -0,0 +1,29 @@
+
+fork(warnmod, {env: {NODE_REDIRECT_WARNINGS: warnpath}})
+  .on('exit', common.mustCall(() => {
+    fs.stat(warnpath, common.mustCall((err, stat) => {
```
Same here. Seems like all of lines 21-28 can be replaced with a single assertion on the results of `readFileSync()`.

Trott replied:
(By the way, these are nits. I'll go back and edit them to indicate that. You can ignore them if you choose. It can always be refactored by someone else in a subsequent PR if it bothers anyone enough. But, you know, if you like to tie up loose ends, there's a possible loose end to tie.)

jasnell replied:
Fixed

jasnell commented:
@Trott @thefourtheye ... LGTY now?

Trott commented:
> @Trott @thefourtheye ... LGTY now?

I haven't reviewed the whole change set, but I have no objections to what I have reviewed.

trevnorris commented:
@jasnell Just noticed there's a code for documentation-only deprecations. Are documentation-only deprecations assigned ids so there's an allocated id when the runtime-deprecation is put into place?

jasnell commented:
Yes, that's it exactly.

sam-github commented on doc/api/_toc.md:
```diff
@@ -14,6 +14,7 @@
 * [Console](console.html)
 * [Crypto](crypto.html)
 * [Debugger](debugger.html)
+* [Deprecated APIs](deprecations.html)
```
nit: maybe use the same word in the file name and the section name? I.e. `deprecated.md/html` or even `deprecated-apis.md/html`?

sam-github commented on doc/api/cli.md:
```diff
@@ -121,6 +121,13 @@ added: v6.0.0
 
 Print stack traces for process warnings (including deprecations).
 
+### `--redirect-warnings=file`
```
I'm not terribly convinced this is useful enough to justify.

1. It can't be triggered by env var, so its hard to configure in the cloud, or as a local user policy for people who just don't care and don't want to see the warnings
2. warnings are emitted as events on process, specifically so this kindof handling can be implemented by the user, so why are we doing it for them?
3. why just warnings? what about an option to redirect stack traces on uncaught errors or exceptions?
4. why not just redirect stderr to a file?

sam-github replied:
Also, docs don't say whether file is created or not, or whether it is appended to, or replaces the file.

jasnell replied:
The idea for this came up during the collaborator summit and was specifically around finding a more intelligent way to handle deprecation warnings that are emitted when using tools like npm. When doing installs, it still may be necessary to capture `stderr` output as normal, but allow warnings to be deferred, captured, and processed in other ways.

This can be configured by environment variable. The envvar is included in this PR.

The file is created if it does not already exist.

sam-github replied:
Sorry, missed the env var version, with that I'm only `-0`.

I was present in that conversation, and I don't think this addresses the experience of node users.

When doing npm installs, the only code that runs is npm's, so its not clear to me why we would want to silence the deprecation warnings. npm should just fix any of their deps to not cause these warnings, they are responsible citizens, they do that.

When using modules installed by npm, then I guess we don't care, usually, and can't do anything about them as a consumer of the module (other than report bugs). But, this redirect doesn't distringuish between deprecation warnings from our dependencies, and deprecations from out code itself, and node can't distinguish, it doesn't really know.

I think in the future it will look like an ad-hoc hack to deal with a more general problem, and we'll have to keep it around for a long time until we deprecate it.

That said, I am not prepared to argue more strongly against it. If noone else chimes in to object, its OK by me.

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+* Runtime
+
+A Documentation-only deprecation is one that is expressed only within the
+Node.js API docs. These generate no side-effects while running a Node.js
```
"while running Node.js." would be shorter, and no less clear

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Node.js API docs. These generate no side-effects while running a Node.js
+process.
+
+A Runtime deprecation will, by default, generate a process warning that will
```
"process warning" as opposed to some other kind of warning? Or do you mean, "emit a `'warning'` event on process"?

jasnell replied:
a process warning is the generic term applied to anything emitted using `process.emitWarning()` / `process.on('warning')`

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Runtime
+
+The `_linklist` module is deprecated. Please use a userland alternative.
```
I think "userland" is too much of linux kernel slang to be useful to node readers. Maybe "Please use a data structures package or implement a linked list directly.", or ref to npmjs.org, or something.

jasnell replied:
For this particular PR, I would rather not delve into the specifics of each individual deprecation description. I'd rather save that for subsequent PRs. Most of these I pulled directly from the warning messages printed to the console

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Documentation-only
+
+The `CryptoStream.prototype.readyState` property is deprecated and should not
+be used.
```
No alternative?

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Documentation-only
+
+The `Buffer()` function and `new Buffer()` constructor are deprecated due to
+API usability issues that can potentially lead to accidental security issues.
```
remove "accidental" (as opposed to deliberate security vulns?)

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Runtime
+
+[`Domain.dispose()`][] is deprecated. Recover from failed I/O actions
+explicitly via error event handlers set on the domain instead.
```
"Or just don't use domains, since the are doc deprecated in their entirety"?

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Runtime
+
+Calling an asynchronous function without callback is deprecated.
```
"a callback"

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Runtime
+
+The `Intl.v8BreakIterator` is deprecated and will be removed or replaced soon.
```
replacement?

targos replied:
The replacement is currently a [stage 2 proposal](https://github.com/tc39/proposal-intl-segmenter)

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Unhandled promise rejections are deprecated. In the future, promise rejections
+that are not handled will terminate the Node.js process with a non-zero exit
+code.
```
definition of unhandled?

addaleax replied:
We probably don’t need to go into so much detail here, a link to the `unhandledRejection` docs should be enough

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Runtime
+
+In certain cases, `require('.')` may resolve outside the package directory.
```
ref to PR or definition of "certain cases"?

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Runtime
+
+The `REPLServer.prototype.convertToContext()` API is deprecated and should
+not be used.
```
replacement?

addaleax replied:
This one has no replacement, people would need to write it themselves if it turns out anybody actually needs it.

sam-github replied:
Just to be clear, I'm not asking for my own personal interest, but because I believe the docs should state what the replacement is, link to more info, or state why there is no replacement. "believed to be useless and unused, write it yourself" is an acceptable reason to doc, IMO.

jasnell replied:
While I agree that the docs can be better, I want to avoid going down the rabbit hole with each of the deprecation descriptions in this PR. Those can be fixed up in subsequent PRs

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Documentation-only
+
+The [`ecdh.setPublicKey()`][] method is now deprecated as its inclusion in the
+API is not useful.
```
no replacement mentioned, but clear statement as to why, +1

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Documentation-only
+
+The [`domain`][] module is deprecated and should not be used.
```
link to the topic describin why?

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Documentation-only
+
+The [`fs.lchmod(path, mode, callback)`][] API has been deprecated.
```
because it can't be supported on most platforms?

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Documentation-only
+
+The [`require.extensions`][] property has been deprecated.
```
people ask about replacements

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+Type: Documentation-only
+
+The [`punycode`][] module has been deprecated. Please use a userland alternative
+instead.
```
why not ", it is now available from npmjs.org, https://www.npmjs.com/package/punycode"

sam-github commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,574 @@
+
+Type: Documentation-only
+
+The [`util.isDate()`][] API has been deprecated.
```
these are all replaced with `typeof X === blah`

sam-github commented on doc/api/process.md:
```diff
@@ -706,14 +706,15 @@ console.log(process.env.test);
-  for the warning. Default: `Warning`.
+* `type` {String} When `warning` is a String, `type` is the name to use
+  for the *type* of warning being emitted. Default: `Warning`.
+* `code` {String} A unique identifier for the warning instance being emitted.
```
I think "number" when I think "code", just me? This seems like an "id", its even described as an identifier in the docs.

jasnell replied:
Use of `code` to contain a string is already well established (https://nodejs.org/dist/latest-v7.x/docs/api/errors.html#errors_error_code) and is consistent with existing practice in core.

sam-github commented on doc/api/process.md:
````diff
@@ -706,14 +706,15 @@ console.log(process.env.test);
 ```
 
-## process.emitWarning(warning[, name][, ctor])
+## process.emitWarning(warning[, type[, code]][, ctor])
````
I think this is basically just intended for internal use, so doesn't need a nice API, and changing the call signature would cause a fair amount of internal churn.

Honestly, I don't think it should be an API at all, maybe we can deprecate it? Seriously, why do we want to expose this to users of node? We expect them to be using it to emit their own warnings?

addaleax replied:
> Honestly, I don't think it should be an API at all, maybe we can deprecate it? Seriously, why do we want to expose this to users of node? We expect them to be using it to emit their own warnings?

Fwiw it was deliberately added as an outward-facing API in c6656db352973d6aea24cb1a3c76adf042b25446 (released in v6.0.0) with that intention, yes. I don’t mind it being there.

jasnell replied:
It was exposed this way because we had users asking to be able to use it also. I'm definitely -1 on deprecating it as I have found it to be quite useful

sam-github replied:
OK, then I agree with @silverwind, its become an ugly API, with three strings in a row, 2 of which are optional, and remembering the arg order isn't reasonably possible. @addaleax the problem with APIs is they rot with time, becoming harder and harder to use and support, like this one is now, which is why I am generally in favour of not adding unnecessary APIs, even if when they are useful. I guess given the alternative of adding a second options-based call pattern, and supporting the old 3 argument call for backwards compatibility, and documenting both of them, or just adding another optional string arg, better to do the latter. If we ever add a 4th string, though, it will no longer be reasonable.

silverwind replied:
@sam-github My thoughts exactly. Supporting two call signatures is easy and enables easier future additons.

jasnell replied:
The API *already* supports two call signatures. The first argument can be passed in as a pre-constructed `Error` object with all of the relevant properties set:

```js
const warning = new Error('This is a warning');
warning.name = 'CustomWarning';
warning.code = 'ABC123';
process.emitWarning(warning);
```

is equivalent to:

```js
process.emitWarning('This is a warning', 'CustomWarning', 'ABC123');
```

Having an options object when we can already pass in an `Error` object would be rather pointless, IMO

sam-github replied:
That Error behaviour isn't well documented. I read carefully, and I see now that its implied by by the "If warning is passed as an Error object, it will be passed through .... unmodified" sentence.

I think the docs could be clearer, but that's an aside from this PR.

jasnell replied:
It's not just implied, it's explicitly stated in the arguments list that `warning` can be a string or an `Error`, and an example of passing an `Error` is shown immediately following the `If warning is passed as an Error object...` sentence. Immediately following the example is a clear statement that a `TypeError` will be thrown is `warning` is anything other than a string or `Error` object. The docs seem pretty explicit about it.

jasnell commented:
Fixed up some bits based on the feedback

jasnell commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,573 @@
+Node.js utilizes two kinds of Deprecations:
+
+* Documentation-only
+* Runtime
```
Looking at it, this could certainly be made clearer and I'll fix that.. but, the idea is that Runtime deprecation *also* includes a deprecation notice in the docs.

jasnell commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,573 @@
+this document will be updated with information relevant to the decision.
+*However, the deprecation identifier will not be modified*.
+
+## List of Deprecated APIs
```
They are in search-result order ;-) ... I assigned the deprecation codes based on the order they appeared in the search results when I was looking for existing deprecated APIs. For this initial set, that's likely good enough. Going forward they would naturally be in chronological order.

Fishrock123 replied:
I just feel people are going to ask us why until the end of time lol :S

jasnell commented on doc/api/cli.md:
```diff
@@ -121,6 +121,16 @@ added: v6.0.0
+added: REPLACEME
+-->
+
+Write process warnings to the given file instead of printing to stderr. The
```
A filepath only (absolute or relative). We could theoretically allow for an fd number but not sure if making it more complicated is worth it.... not out of the question tho.

jasnell commented:
@nodejs/ctc ... ping again.

jasnell commented:
Rebased...

targos commented on lib/internal/process/warning.js:
```diff
@@ -1,9 +1,76 @@
+  };
+}
+
+function acquireFd(cb) {
```
shouldn't it have a third condition for the case where `acquiringFd` is false and `fd` exists? I think it should call `cb(null, fd)` immediately. Otherwise, almost each call to `output` will open a new fd.

targos commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,573 @@
+<a id="DEP0014"></a>
+### DEP0014: fs.read legacy String interface
+
+Type: Runtime
```
This legacy interface has been removed. Is there another type to reflect that?

jasnell replied:
In the new deprecation policy a new "end-of-life" status is introduced. We should use that here.

targos commented on doc/api/deprecations.md:
```diff
@@ -0,0 +1,573 @@
+<a id="DEP0015"></a>
+### DEP0015: fs.readSync legacy String interface
+
+Type: Runtime
```
ditto

rvagg commented on lib/internal/process.js:
```diff
@@ -128,7 +128,8 @@ function setupConfig(_source) {
       return Reflect.construct(oldV8BreakIterator, arguments);
-    }, 'Intl.v8BreakIterator is deprecated and will be removed soon.');
+    }, 'Intl.v8BreakIterator is deprecated and will be removed soon.',
+                                                   'DEP0017');
```
what's going on with the indenting here? seems wrong to me cause it's lining up with the wrong line.

jasnell commented:
I noticed the weirdness with three indenting either. Lint wouldn't pass
without it. Wasn't too happy but not unhappy enough to do anything about it
yet.

joshgav commented:
I'd like to reserve `DEP0062` for #10970.

jasnell commented:
We'll need to figure out a process for that. Reserving specific numbers may be problematic going forward, especially in the theoretical situation where we have more than one pending deprecation and it's not clear in what order they will land. However, if you'd like to open a pr to add that it would be fine

jasnell commented:
I will be working to address the feedback from @targos and will look to land this on Monday

jasnell commented:
@targos updated! PTAL
CI: https://ci.nodejs.org/job/node-test-pull-request/6084/

jasnell commented:
Landed in 5e1f32f..5de3cf0

sam-github commented:
@jasnell do you plan to backport this? Its a useful feature, and it appears to not be necessarily semver-major:

> Note: This PR is currently semver-major because it adds improved type checking to the input on process.emitWarning() to ensure that both the warning name and the identifier code are strings.
If this PR is backported to older Node.js release streams, the backport can be landed as a semver-minor if and only if the additional type checking on the name argument is removed.

Its blocking later PRs that touch nearby code even though they don't depend on this feature, like #11051, and its follow ons.

I can backport if you don't have time, but would like it in 7.x.

jasnell commented:
I backported it to 4.x today and opened a PR. It should be fairly trivial to backport to 7.x and 6.x. I wouldn't be able to get to it until next week so if you'd like to go for it, be my guest. 

btw... I would also include the new `NODE_NO_WARNINGS` environment variable if that has not already been backported

sam-github commented:
I backported 03e89b3ff298c63e9620 to 6.x, the other commits look semver-major to me, as they add deprecation code and change the output format. Do I misunderstand?
